### ISSUE TITLE
Huge CPU, GPU output diff after updates to fully_connected GpuDelegate gl kernel.

### ISSUE THREAD

**Issue by prodia4** [tensorflow/tensorflow#42294]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
No

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Android

- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
Tested on Pixel2

- TensorFlow installed from (source or binary):
Built from source

- TensorFlow version (use command below):
v2.3.0

- Python version:
3.7

- Bazel version (if compiling from source):
3.1.0

- GCC/Compiler version (if compiling from source):
7.5.0

- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
There seems to be a bug in tensorflow/lite/delegates/gpu/gl/kernels/fully_connected.cc that was first updated in [this commit](https://github.com/tensorflow/tensorflow/commit/7c7d924821a8b1b20433c2f3f484bbd409873a84#diff-a1d24db3e0ad4103e389ffed433cf470R82).

The test itself in tensorflow/lite/delegates/gpu/gl/kernels/fully_connected_test.cc passes but when running
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/tools/evaluation/tasks/inference_diff on a simple custom model, the avg_error between the CPU and the GPU delegate piles up and leads to completely different results.

**Describe the expected behavior**
The avg_error between the CPU and the GPU delegate should stay the same even after the updates to fully_connected code, and it should not produce weird outputs.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

1. Prepare `run_eval` in v2.3.0
```
bazel build -c opt \
  --config=android_arm64 \
  //tensorflow/lite/tools/evaluation/tasks/inference_diff:run_eval
cp bazel-bin/tensorflow/lite/tools/evaluation/tasks/inference_diff/run_eval ./run_eval_1
```

2. Change the fully_connected shader implementation to the one prior to commit `7c7d9248`
The simple implementation without the usage of shared variables, memoryBarrierShared, etc.

3. Prepare `run_eval` after second step
```
bazel build -c opt \
  --config=android_arm64 \
  //tensorflow/lite/tools/evaluation/tasks/inference_diff:run_eval
cp bazel-bin/tensorflow/lite/tools/evaluation/tasks/inference_diff/run_eval ./run_eval_2
```

4. Connect android device and push the tflite models below, run_eval_1 and run_eval_2 to device
```
adb push [] /data/local/tmp
```
[simple.zip](https://github.com/tensorflow/tensorflow/files/5066647/simple.zip)
The models in the zip file are submodels from a custom model that was sliced from the input layer to deeper layers accumulatively, so that running `run_eval` on them would show how the error changes in the subsequent layers.

5. Run both eval scripts on the models and save the output
```
# For example
adb shell find /data/local/tmp/simple -type f -name "*.tflite" | sort -V | xargs -L 1 -I{} sh -c "echo {}; adb shell /data/local/tmp/run_eval_1 --model_file={} --delegate=gpu --gpu_precision_loss_allowed=false --num_runs=5" > eval_1.txt
```

6. Compare output from both eval results
```
[Output from run_eval_1]

/data/local/tmp/simple/0_layer.tflite (input)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=350.6(us), std_dev=182(us)
Test run latency: avg=7788.33(us), std_dev=1361(us)
OutputDiff[0]: avg_error=0, std_dev=0

/data/local/tmp/simple/1_layer.tflite (conv2d)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=51505.2(us), std_dev=19231(us)
Test run latency: avg=76189.3(us), std_dev=21566(us)
OutputDiff[0]: avg_error=5.65735e-08, std_dev=1.23668e-10

/data/local/tmp/simple/2_layer.tflite (conv2d)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=156689(us), std_dev=34476(us)
Test run latency: avg=91240.9(us), std_dev=20015(us)
OutputDiff[0]: avg_error=2.88166e-07, std_dev=4.7596e-10

/data/local/tmp/simple/3_layer.tflite (average_pooling2d)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=154217(us), std_dev=49214(us)
Test run latency: avg=63426.3(us), std_dev=3858(us)
OutputDiff[0]: avg_error=2.06003e-07, std_dev=5.04424e-10

/data/local/tmp/simple/4_layer.tflite (average_pooling2d)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=154598(us), std_dev=56655(us)
Test run latency: avg=43006.3(us), std_dev=8692(us)
OutputDiff[0]: avg_error=8.01869e-08, std_dev=8.87667e-09

/data/local/tmp/simple/5_layer.tflite (dense)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=156916(us), std_dev=54132(us)
Test run latency: avg=36966.8(us), std_dev=1220(us)
OutputDiff[0]: avg_error=0, std_dev=0

/data/local/tmp/simple/6_layer.tflite (dense)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=152091(us), std_dev=53248(us)
Test run latency: avg=36698(us), std_dev=1180(us)
OutputDiff[0]: avg_error=0.00405404, std_dev=8.56487e-06

/data/local/tmp/simple/7_layer.tflite (reshape)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=157181(us), std_dev=52575(us)
Test run latency: avg=39436.9(us), std_dev=2816(us)
OutputDiff[0]: avg_error=0.00405404, std_dev=8.56487e-06

/data/local/tmp/simple/8_layer.tflite (mul)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=151318(us), std_dev=51893(us)
Test run latency: avg=47089.9(us), std_dev=3074(us)
OutputDiff[0]: avg_error=0.00415553, std_dev=1.17143e-05

/data/local/tmp/simple/9_layer.tflite (conv2d)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=176098(us), std_dev=49177(us)
Test run latency: avg=51363.7(us), std_dev=933(us)
OutputDiff[0]: avg_error=0.0141088, std_dev=9.8753e-05

/data/local/tmp/simple/10_layer.tflite (conv2d)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=194176(us), std_dev=42305(us)
Test run latency: avg=56783.7(us), std_dev=942(us)
OutputDiff[0]: avg_error=0.116762, std_dev=0.00102424
```

```
[Output from run_eval_2]

/data/local/tmp/simple/0_layer.tflite (input)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=404.067(us), std_dev=291(us)
Test run latency: avg=9060.07(us), std_dev=2928(us)
OutputDiff[0]: avg_error=0, std_dev=0

/data/local/tmp/simple/1_layer.tflite (conv2d)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=49690.5(us), std_dev=18876(us)
Test run latency: avg=76622.3(us), std_dev=16760(us)
OutputDiff[0]: avg_error=5.65735e-08, std_dev=1.23668e-10

/data/local/tmp/simple/2_layer.tflite (conv2d)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=154928(us), std_dev=37358(us)
Test run latency: avg=91402.5(us), std_dev=20801(us)
OutputDiff[0]: avg_error=2.88166e-07, std_dev=4.7596e-10

/data/local/tmp/simple/3_layer.tflite (average_pooling2d)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=156680(us), std_dev=51722(us)
Test run latency: avg=63421.1(us), std_dev=5300(us)
OutputDiff[0]: avg_error=2.06003e-07, std_dev=5.04424e-10

/data/local/tmp/simple/4_layer.tflite (average_pooling2d)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=168827(us), std_dev=51702(us)
Test run latency: avg=42583.5(us), std_dev=9370(us)
OutputDiff[0]: avg_error=8.01869e-08, std_dev=8.87667e-09

/data/local/tmp/simple/5_layer.tflite (dense)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=153892(us), std_dev=54096(us)
Test run latency: avg=38346.1(us), std_dev=2488(us)
OutputDiff[0]: avg_error=0, std_dev=0

/data/local/tmp/simple/6_layer.tflite (dense)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=158208(us), std_dev=54326(us)
Test run latency: avg=38795.7(us), std_dev=2783(us)
OutputDiff[0]: avg_error=2.78233e-08, std_dev=0

/data/local/tmp/simple/7_layer.tflite (reshape)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=159020(us), std_dev=56125(us)
Test run latency: avg=36865.9(us), std_dev=1231(us)
OutputDiff[0]: avg_error=2.78233e-08, std_dev=0

/data/local/tmp/simple/8_layer.tflite (mul)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=155681(us), std_dev=51329(us)
Test run latency: avg=46260.3(us), std_dev=2496(us)
OutputDiff[0]: avg_error=3.92487e-08, std_dev=9.90499e-11

/data/local/tmp/simple/9_layer.tflite (conv2d)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=192473(us), std_dev=49285(us)
Test run latency: avg=68372(us), std_dev=8581(us)
OutputDiff[0]: avg_error=1.127e-07, std_dev=2.97108e-10

/data/local/tmp/simple/10_layer.tflite (conv2d)
GPU delegate is created.
Num evaluation runs: 5
Reference run latency: avg=184917(us), std_dev=50705(us)
Test run latency: avg=57926.3(us), std_dev=4106(us)
OutputDiff[0]: avg_error=6.8757e-07, std_dev=1.08467e-09
```

in `run_eval_1`, the `avg_error` rises high after the dense layers, when compared to `run_eval_2`.

This leads to a weird output in the custom model that I'm using in an Android app. When I rollback the changes in the `fully_connected` code and build the aar myself, it works as expected and I can benefit the performance enhancements on devices that support OpenCL.

Could you please have a look at it?

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by impjdi** [comment]
> When I rollback the changes in the `fully_connected` code and build the aar myself, it works as expected and I can benefit the performance enhancements on devices that support OpenCL.

Can you clarify which commit breaks it?  Should make it much easier if you could pinpoint the problematic commit.

**Comment by prodia4** [comment]
The commit was linked above, but let me paste it again here for convenience.

[https://github.com/tensorflow/tensorflow/commit/7c7d924821a8b1b20433c2f3f484bbd409873a84](https://github.com/tensorflow/tensorflow/commit/7c7d924821a8b1b20433c2f3f484bbd409873a84)

**Comment by impjdi** [comment]
Must've missed that.  Thank you, Kevin!

**Comment by impjdi** [comment]
Wow, that's a commit from a year ago.  Either nobody is using FC, or you're hitting a weird corner case.  We'll look into that.

**Comment by impjdi** [comment]
I got in touch with the engineer who worked on that commit.  He told me that the implementation had a flaw but got fixed with commit `522aa755f9e4cdb37ab8d80ac252df60553c6439` from Sep 12, 2019.  Can you sync to that?

**Comment by prodia4** [comment]
The comparison was done between 

A. The code in tag v2.3.0, which includes the code changes you mentioned in `522aa755f9e4cdb37ab8d80ac252df60553c6439`

and 

B. Just the `fully_connected.cc` replaced by its former implementation, prior to commit `7c7d924821a8b1b20433c2f3f484bbd409873a84`

So I don't think commit `522aa755f9e4cdb37ab8d80ac252df60553c6439` fixed the issue.
